File: apps/daemon/crates/falco/src/courier.rs
//! Main courier loop orchestration.

use crate::config::FalcoConfig;
use crate::daemon_client::{connect_with_retry, DaemonClient};
use crate::error::{FalcoError, FalcoResult};
use crate::protocol::Decision;
use crate::redis_consumer::RedisConsumer;
use std::time::Duration;
use tracing::{debug, error, info, warn};
use uuid::Uuid;

/// The main Falco courier.
///
/// Orchestrates the flow of commands from Redis to the daemon.
pub struct Courier {
    config: FalcoConfig,
    redis: RedisConsumer,
    daemon: Option<DaemonClient>,
}

impl Courier {
    /// Create a new Courier.
    pub async fn new(config: FalcoConfig) -> FalcoResult<Self> {
        let redis = RedisConsumer::connect(config.clone()).await?;

        Ok(Self {
            config,
            redis,
            daemon: None,
        })
    }

    /// Run the main courier loop.
    ///
    /// This loop:
    /// 1. Reads one message from Redis (XREADGROUP COUNT=1)
    /// 2. Generates a command_id UUID
    /// 3. Connects to daemon if needed
    /// 4. Sends CommandFrame to daemon
    /// 5. Waits for DaemonDecisionFrame (with timeout)
    /// 6. ACKs Redis if instructed or on timeout
    /// 7. Loops
    pub async fn run(&mut self) -> FalcoResult<()> {
        info!(
            device_id = %self.config.device_id,
            stream = %self.config.stream_key(),
            consumer = %self.config.consumer_name,
            "Starting Falco courier loop"
        );

        loop {
            if let Err(e) = self.process_one().await {
                error!(error = %e, "Error processing message");

                // On certain errors, wait before retrying
                match &e {
                    FalcoError::Redis(_) => {
                        warn!("Redis error, attempting to reconnect...");
                        tokio::time::sleep(Duration::from_secs(1)).await;
                        if let Err(reconnect_err) = self.redis.reconnect().await {
                            error!(error = %reconnect_err, "Failed to reconnect to Redis");
                            tokio::time::sleep(Duration::from_secs(5)).await;
                        }
                    }
                    FalcoError::DaemonConnection(_) => {
                        warn!("Daemon connection error, will retry on next message");
                        self.daemon = None;
                        tokio::time::sleep(Duration::from_millis(500)).await;
                    }
                    _ => {
                        // For other errors, brief pause before continuing
                        tokio::time::sleep(Duration::from_millis(100)).await;
                    }
                }
            }
        }
    }

    /// Process one message from Redis.
    async fn process_one(&mut self) -> FalcoResult<()> {
        // Read the next message from Redis
        let message = match self.redis.read_next().await? {
            Some(msg) => msg,
            None => {
                // Block timeout expired, no messages available
                debug!("No messages available, continuing to poll...");
                return Ok(());
            }
        };

        info!(
            message_id = %message.message_id,
            payload_len = message.encrypted_payload.len(),
            "Processing message"
        );

        // Generate a command ID for correlation
        let command_id = Uuid::new_v4();

        // Ensure daemon connection
        self.ensure_daemon_connection().await?;

        let daemon = self.daemon.as_mut().ok_or_else(|| {
            FalcoError::DaemonConnection("No daemon connection".to_string())
        })?;

        // Send to daemon and wait for decision
        let decision = match daemon
            .send_and_wait(
                command_id,
                message.encrypted_payload.clone(),
                self.config.daemon_timeout,
            )
            .await
        {
            Ok((decision, _result)) => decision,
            Err(FalcoError::Timeout(secs)) => {
                // Fail-open: ACK on timeout to prevent infinite PEL growth
                warn!(
                    message_id = %message.message_id,
                    command_id = %command_id,
                    timeout_secs = secs,
                    "Daemon timeout, applying fail-open ACK"
                );
                Decision::AckRedis
            }
            Err(e) => {
                // Other errors: don't ACK, let Redis redeliver
                error!(
                    message_id = %message.message_id,
                    command_id = %command_id,
                    error = %e,
                    "Error communicating with daemon, leaving message in PEL"
                );
                self.daemon = None; // Force reconnect on next attempt
                return Err(e);
            }
        };

        // Apply the decision
        match decision {
            Decision::AckRedis => {
                self.redis.ack(&message.message_id).await?;
                info!(
                    message_id = %message.message_id,
                    command_id = %command_id,
                    "Message acknowledged"
                );
            }
            Decision::DoNotAck => {
                info!(
                    message_id = %message.message_id,
                    command_id = %command_id,
                    "Message left in PEL (daemon requested no ACK)"
                );
                // Message stays in PEL for redelivery
                // This could be because the daemon wants to retry later
            }
        }

        Ok(())
    }

    /// Ensure we have a connection to the daemon.
    async fn ensure_daemon_connection(&mut self) -> FalcoResult<()> {
        if self.daemon.is_some() {
            return Ok(());
        }

        info!(
            socket = %self.config.socket_path.display(),
            "Connecting to daemon..."
        );

        let client = connect_with_retry(
            &self.config.socket_path,
            3,                            // max retries
            Duration::from_millis(500),   // retry delay
        )
        .await?;

        self.daemon = Some(client);
        info!("Connected to daemon");

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Integration tests would require a running Redis and daemon
    // Unit tests for Courier are limited since it orchestrates external services

    #[test]
    fn test_config_stream_key() {
        let config = FalcoConfig::new("test-device".to_string()).unwrap();
        assert_eq!(config.stream_key(), "remote:commands:test-device");
    }
}

File: packages/daemon-nagato/courier/courier.go
// Package courier implements the main orchestration loop for Nagato.
//
// The courier is responsible for:
// 1. Receiving messages from Ably (one at a time)
// 2. Forwarding encrypted payloads to the daemon
// 3. Waiting for the daemon's decision
// 4. Handling timeout fail-open behavior
//
// # Key Invariants
//
//   - One-in-flight: Only one message is processed at a time
//   - ACK-gated: Messages are only considered processed when daemon responds OR timeout occurs (fail-open)
//   - Content-agnostic: Payloads are never inspected or modified
//   - Crash-safe: No persistent state; recovery handled by Ably message redelivery
package courier

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"sync"
	"time"

	"github.com/google/uuid"
	"go.uber.org/zap"

	"github.com/unbound-computer/daemon-nagato/client"
	"github.com/unbound-computer/daemon-nagato/config"
	"github.com/unbound-computer/daemon-nagato/consumer"
	"github.com/unbound-computer/daemon-nagato/protocol"
)

const (
	// Error recovery delays
	ConnectionErrorDelay = time.Second
	ProtocolErrorDelay   = 100 * time.Millisecond

	// Event emitted by the machine after command handling.
	RemoteCommandAckEvent = "remote.command.ack.v1"
)

const (
	AckStatusAccepted = "accepted"
	AckStatusRejected = "rejected"
	AckStatusTimeout  = "timeout"
)

var (
	ErrShutdown = errors.New("courier shutdown")
)

type commandAckPayload struct {
	SchemaVersion int    `json:"schema_version"`
	CommandID     string `json:"command_id"`
	Status        string `json:"status"`
	CreatedAtMS   int64  `json:"created_at_ms"`
	ResultB64     string `json:"result_b64,omitempty"`
}

// Courier orchestrates message flow between Ably and the daemon.
type Courier struct {
	cfg      *config.Config
	consumer *consumer.Consumer
	daemon   *client.Client
	logger   *zap.Logger

	mu       sync.Mutex
	running  bool
	cancelFn context.CancelFunc
}

// New creates a new Courier instance.
func New(cfg *config.Config, logger *zap.Logger) (*Courier, error) {
	if logger == nil {
		logger = zap.NewNop()
	}

	// Create Ably consumer
	cons, err := consumer.New(consumer.Options{
		AblySocketPath: cfg.AblySocketPath,
		ChannelName:    cfg.ChannelName,
		EventName:      cfg.EventName,
		SubscriptionID: cfg.ConsumerName,
		Logger:         logger.Named("consumer"),
		BufferSize:     1, // One-in-flight
	})
	if err != nil {
		return nil, err
	}

	// Create daemon client
	daemonClient := client.NewClient(
		cfg.SocketPath,
		cfg.DaemonTimeout,
		logger.Named("daemon"),
	)

	return &Courier{
		cfg:      cfg,
		consumer: cons,
		daemon:   daemonClient,
		logger:   logger,
	}, nil
}

// Run starts the courier loop.
// This blocks until the context is cancelled or an unrecoverable error occurs.
func (c *Courier) Run(ctx context.Context) error {
	c.mu.Lock()
	if c.running {
		c.mu.Unlock()
		return errors.New("courier already running")
	}
	c.running = true

	ctx, cancel := context.WithCancel(ctx)
	c.cancelFn = cancel
	c.mu.Unlock()

	defer func() {
		c.mu.Lock()
		c.running = false
		c.cancelFn = nil
		c.mu.Unlock()
	}()

	c.logger.Info("starting courier",
		zap.String("device_id", c.cfg.DeviceID),
		zap.String("channel", c.cfg.ChannelName),
		zap.String("event", c.cfg.EventName),
		zap.String("socket", c.cfg.SocketPath),
		zap.String("consumer", c.cfg.ConsumerName),
		zap.Duration("timeout", c.cfg.DaemonTimeout),
	)

	// Connect to Ably
	if err := c.consumer.Connect(ctx); err != nil {
		return err
	}
	defer c.consumer.Close()

	// Connect to daemon
	if err := c.daemon.Connect(ctx); err != nil {
		c.logger.Warn("initial daemon connection failed, will retry",
			zap.Error(err),
		)
		// Don't fail - we'll retry on each message
	}
	defer c.daemon.Close()

	c.logger.Info("courier started, entering main loop")

	// Main loop
	return c.runLoop(ctx)
}

// runLoop is the main message processing loop.
func (c *Courier) runLoop(ctx context.Context) error {
	messages := c.consumer.Receive()

	for {
		select {
		case <-ctx.Done():
			c.logger.Info("courier shutting down",
				zap.Error(ctx.Err()),
			)
			return ErrShutdown

		case msg, ok := <-messages:
			if !ok {
				c.logger.Info("message channel closed")
				return ErrShutdown
			}

			if err := c.processMessage(ctx, msg); err != nil {
				c.logger.Error("error processing message",
					zap.String("message_id", msg.ID),
					zap.Error(err),
				)

				// Apply error recovery delay
				select {
				case <-ctx.Done():
					return ErrShutdown
				case <-time.After(ProtocolErrorDelay):
				}
			}
		}
	}
}

// processMessage handles a single message from Ably.
func (c *Courier) processMessage(ctx context.Context, msg *consumer.Message) error {
	c.logger.Debug("processing message",
		zap.String("message_id", msg.ID),
		zap.Int("payload_len", len(msg.Payload)),
	)

	// Ensure daemon connection
	if !c.daemon.IsConnected() {
		c.logger.Debug("reconnecting to daemon")
		if err := c.daemon.Connect(ctx); err != nil {
			c.logger.Error("failed to connect to daemon",
				zap.Error(err),
			)
			// Message will be redelivered by Ably on reconnect
			return err
		}
	}

	// Send to daemon and wait for decision
	commandID, response, err := c.daemon.SendAndWait(ctx, msg.Payload)
	if err != nil {
		// Check if it's a timeout (fail-open)
		if errors.Is(err, client.ErrTimeout) {
			c.logger.Warn("daemon timeout, applying fail-open",
				zap.String("message_id", msg.ID),
				zap.Duration("timeout", c.cfg.DaemonTimeout),
			)
			c.publishAck(ctx, commandID, msg.ID, AckStatusTimeout, nil)
			// Fail-open: consider message processed to prevent blocking
			// The daemon may have processed it but failed to respond
			return nil
		}

		// Other errors
		c.logger.Error("daemon communication error",
			zap.String("message_id", msg.ID),
			zap.Error(err),
		)
		return err
	}

	// Log daemon's decision
	switch response.Decision {
	case protocol.AckMessage:
		c.logger.Info("daemon processed message successfully",
			zap.String("message_id", msg.ID),
		)
		c.publishAck(ctx, response.CommandID, msg.ID, AckStatusAccepted, response.Result)

	case protocol.DoNotAck:
		c.logger.Warn("daemon rejected message",
			zap.String("message_id", msg.ID),
		)
		c.publishAck(ctx, response.CommandID, msg.ID, AckStatusRejected, response.Result)
		// Message may be redelivered depending on Ably configuration

	default:
		c.logger.Error("unexpected decision",
			zap.String("message_id", msg.ID),
			zap.Uint8("decision", uint8(response.Decision)),
		)
	}

	return nil
}

func (c *Courier) publishAck(
	ctx context.Context,
	commandID uuid.UUID,
	ablyMessageID string,
	status string,
	result []byte,
) {
	if commandID == uuid.Nil {
		c.logger.Warn("skipping command ack publish due to missing command_id",
			zap.String("ably_message_id", ablyMessageID),
			zap.String("status", status),
		)
		return
	}

	payload := commandAckPayload{
		SchemaVersion: 1,
		CommandID:     commandID.String(),
		Status:        status,
		CreatedAtMS:   time.Now().UnixMilli(),
	}
	if len(result) > 0 {
		payload.ResultB64 = base64.StdEncoding.EncodeToString(result)
	}

	encoded, err := json.Marshal(payload)
	if err != nil {
		c.logger.Warn("failed to encode command ack payload",
			zap.String("ably_message_id", ablyMessageID),
			zap.String("command_id", commandID.String()),
			zap.Error(err),
		)
		return
	}

	if err := c.consumer.Publish(ctx, RemoteCommandAckEvent, encoded); err != nil {
		c.logger.Warn("failed to publish command ack",
			zap.String("ably_message_id", ablyMessageID),
			zap.String("command_id", commandID.String()),
			zap.String("status", status),
			zap.Error(err),
		)
	}
}

// Stop gracefully stops the courier.
func (c *Courier) Stop() {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.cancelFn != nil {
		c.logger.Info("stopping courier")
		c.cancelFn()
	}
}

// IsRunning returns true if the courier is currently running.
func (c *Courier) IsRunning() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.running
}
